xcodeproj/internal/target_id.bzl

```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
     """
     return "{} {}".format(label, configuration)
 
+def _longest_common_prefix(labels):
+    if not labels:
+        return None
+
+    # Assume the shortest string is the answer
+    res = labels[0]
+    for s in labels:
+        if len(s) < len(res):
+            res = s
+
+    # Compare one character at a time with the rest,
+    # as soon as any difference is found the result holds the longest common prefix
+    index = 0
+    for c in res.elems():
+        for label in labels:
+            if label[index] != c:
+                return res[:index]
+        index += 1
+
+    # If the above never returns the shortest string is the result
+    return res
+
+def calculate_replacement_label(*, id, replacement_labels):
+    """Calculates a single replacement label for a given id and list of replacement labels
+
+    It considers the list of labels received plus the label contained in the id itself, then it finds
+    the longest common prefix between those
+
+    Args:
+        id: An id that uniquely represents a target (see `get_id`)
+        replacement_labels: A list of replacement labels
+
+    Returns:
+        The longest common prefix between the list of labels and the label contained in the id,
+        if a longest common prefix is not found the label contained in the id is returned
+    """
+    id_label = id.split(" ")[0]  # This assumes the id follows the creation pattern in `get_id`
+
+    res = _longest_common_prefix(
+        [id_label] + ["%s" % label for label in replacement_labels],
+    )
+    if not res:
+        return None
+
+    res = bazel_labels.normalize_label(res)
```

Comment:
No need to normalize here, as you should do it above (since `bazel_labels.normalize_label` basically does `str()` already, with a little more magic).